core/liquidity/v2/sla.go

```diff
@@ -0,0 +1,229 @@
+		}
+
+		s := num.DecimalFromInt64(commitment.s.Nanoseconds())
+		observedEpochLengthD := num.DecimalFromInt64(observedEpochLength.Nanoseconds())
+		timeBookFraction := s.Div(observedEpochLengthD)
+
+		var feePenalty, bondPenalty num.Decimal
+
+		// if LP meets commitment
+		// else LP does not meet commitment
+		if timeBookFraction.LessThan(e.commitmentMinTimeFraction) {
+			feePenalty = num.DecimalOne()
+			bondPenalty = e.calculateBondPenalty(timeBookFraction)
+		} else {
+			feePenalty = e.calculateCurrentFeePenalty(timeBookFraction)
+			bondPenalty = num.DecimalZero()
+		}
+
+		commitment.allEpochsPenalties = append(commitment.allEpochsPenalties, feePenalty)
+		previousPenalties := commitment.allEpochsPenalties[:len(commitment.allEpochsPenalties)-1]
+
+		penaltiesPerParty[party] = &SlaPenalty{
+			Bond: bondPenalty,
+			Fee:  e.calculateHysteresisFeePenalty(feePenalty, previousPenalties),
+		}
+	}
+
+	e.slaPenalties = penaltiesPerParty
+}
+
+func (e *Engine) GetSLAPenalties() map[string]*SlaPenalty {
+	return e.slaPenalties
+}
+
+func (e *Engine) getMidPrice() (*num.Uint, error) {
+	bestBid, err := e.orderBook.GetBestStaticBidPrice()
+	if err != nil {
+		return nil, err
+	}
+
+	bestAsk, err := e.orderBook.GetBestStaticAskPrice()
+	if err != nil {
+		return nil, err
+	}
+
+	two := num.NewUint(2)
+	midPrice := num.UintZero()
+	if !bestBid.IsZero() && !bestAsk.IsZero() {
+		midPrice = midPrice.Div(num.Sum(bestBid, bestAsk), two)
+	}
+
+	return midPrice, nil
+}
+
+func (e *Engine) doesLPMeetsCommitment(party string) bool {
+	lp, ok := e.provisions.Get(party)
+	if !ok {
+		return false
+	}
+
+	one := num.DecimalOne()
+
+	var minPrice, maxPrice num.Decimal
+	if e.auctionState.InAuction() {
+		minPriceFactor := num.Min(e.orderBook.GetLastTradedPrice(), e.orderBook.GetIndicativePrice()).ToDecimal()
+		maxPriceFactor := num.Max(e.orderBook.GetLastTradedPrice(), e.orderBook.GetIndicativePrice()).ToDecimal()
+
+		// (1.0-market.liquidity.priceRange) x min(last trade price, indicative uncrossing price)
+		minPrice = one.Sub(e.priceRange).Mul(minPriceFactor)
+		// (1.0+market.liquidity.priceRange) x max(last trade price, indicative uncrossing price)
+		maxPrice = one.Add(e.priceRange).Mul(maxPriceFactor)
+	} else {
+		mid, err := e.getMidPrice()
+		// if there is no mid price then LP is not meeting their committed volume of notional.
+		if err != nil || mid.IsZero() {
+			return false
+		}
+
+		midD := mid.ToDecimal()
+		// (1.0 - market.liquidity.priceRange) x mid
+		minPrice = one.Sub(e.priceRange).Mul(midD)
+		// (1.0 + market.liquidity.priceRange) x mid
+		maxPrice = one.Add(e.priceRange).Mul(midD)
+	}
+
+	notionalVolume := num.DecimalZero()
+	orders := e.getAllActiveOrders(party)
+
+	for _, o := range orders {
+		price := o.Price.ToDecimal()
+		// this order is in range and does contribute to the volume on notional
+		if price.GreaterThanOrEqual(minPrice) && price.LessThanOrEqual(maxPrice) {
+			notionalVolume = notionalVolume.Add(price)
```

Comment:
should this not be adding up the orders sizes and not the price to get the LP's volume in that price range? i.e
```
notionalVolume.Add(o.TrueRemaining())
```

Reply:
I think you are right.